packages/libs/error-stack/src/ext/stream.rs

```diff
@@ -0,0 +1,420 @@
+    where
+        C: Context,
+        F: FnOnce() -> C;
+}
+
+impl<S> StreamReportExt for S
+where
+    S: Stream,
+    S::Item: ResultExt,
+{
+    #[track_caller]
+    fn attach<A>(self, attachment: A) -> StreamWithAttachment<Self, A>
+    where
+        A: Send + Sync + 'static,
+    {
+        StreamWithAttachment {
+            stream: self,
+            attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_lazy<A, F>(self, attachment: F) -> StreamWithLazyAttachment<Self, F>
+    where
+        A: Send + Sync + 'static,
+        F: FnOnce() -> A,
+    {
+        StreamWithLazyAttachment {
+            stream: self,
+            attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_printable<A>(self, attachment: A) -> StreamWithPrintableAttachment<Self, A>
+    where
+        A: Display + Debug + Send + Sync + 'static,
+    {
+        StreamWithPrintableAttachment {
+            stream: self,
+            attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_printable_lazy<A, F>(
+        self,
+        attachment: F,
+    ) -> StreamWithLazyPrintableAttachment<Self, F>
+    where
+        A: Display + Debug + Send + Sync + 'static,
+        F: FnOnce() -> A,
+    {
+        StreamWithLazyPrintableAttachment {
+            stream: self,
+            attachment,
+        }
+    }
+
+    #[track_caller]
+    fn change_context<C>(self, context: C) -> StreamWithContext<Self, C>
+    where
+        C: Context,
+    {
+        StreamWithContext {
+            stream: self,
+            attachment: context,
+        }
+    }
+
+    #[track_caller]
+    fn change_context_lazy<C, F>(self, context: F) -> StreamWithLazyContext<Self, F>
+    where
+        C: Context,
+        F: FnOnce() -> C,
+    {
+        StreamWithLazyContext {
+            stream: self,
+            attachment: context,
+        }
+    }
+}
+
+#[pin_project]
+pub struct StreamWithAttachment<S, A> {
+    #[pin]
+    stream: S,
+    attachment: A,
+}
+
+impl<S, A> Stream for StreamWithAttachment<S, A>
+where
+    S: Stream,
+    S::Item: ResultExt,
+    A: Clone + Send + Sync + 'static,
+{
+    type Item = S::Item;
+
+    #[track_caller]
+    fn poll_next(
+        self: core::pin::Pin<&mut Self>,
+        cx: &mut core::task::Context<'_>,
+    ) -> core::task::Poll<Option<Self::Item>> {
+        let projected = self.project();
+        let stream = projected.stream;
+
+        match stream.poll_next(cx) {
+            Poll::Ready(data) => Poll::Ready(match data {
+                Some(data) => Some(data.attach(projected.attachment.clone())),
+                None => None,
+            }),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+#[pin_project]
+pub struct StreamWithLazyAttachment<S, F> {
+    #[pin]
+    stream: S,
+    attachment: F,
+}
+
+impl<S, A, F> Stream for StreamWithLazyAttachment<S, F>
+where
+    S: Stream,
+    S::Item: ResultExt,
+    F: Fn() -> A,
+    A: Clone + Copy + Send + Sync + 'static,
```

Comment:
`A` does is not required to be `Clone` (same as in #716). This also applies to other usages in this file.